add experiment to add new site button

roo2 opened the pull request:
#### Changes proposed in this Pull Request

Add experiment to just the add-site button, see pbxNRc-1jD-p2

fixes #60166 

The new experiments platform code uses [react hooks to get experiments](https://github.com/Automattic/wp-calypso/blob/trunk/client/lib/explat/README.md), since we already have a somewhat involved function for getting the onboardingUrl, I created a combined useOnboardingUrl hook which could take all of the neccessary logic into account.

There are other links to create a new site, but I don't believe these need to be updated (I got a bit lost in the weeds trying to do that) I don't think it's necessary to update all links. I had a think about whether it will be neccessary to add new events to use as assignment filters. pbxNRc-1jD-p2#comment-3011

#### Testing instructions

Open the "Switch Site" menu, the "Add New Site" button at the bottom should take you to `/new` following existing logic. If 
Related to #60166

roo2 committed:
add experiment to add new site button

p-jackson commented on client/components/site-selector/add-site.tsx:
```diff
@@ -0,0 +1,35 @@
+	const translate = useTranslate();
+	const dispatch = useDispatch();
+
+	const recordAddNewSite = useCallback( () => {
+		const event = isJetpackCloud()
```
I was trying to think of what to do when `isLoadingExperimentAssignment` returns true. It seems pretty unlikely that the user would be so fast that they could quickly find the "Add new site" button before the assignment was loaded. But if they did, would it be ok to just stop the click from working? But then when they click it again it magically works?

I'm worried `<a href="?ref=calypso-selector">` will reload the page or something.

```suggestion
	const recordAddNewSite = useCallback( ( e: React.MouseEvent< HTMLAnchorElement > ) => {
		if ( ! onboardingUrl ) {
			e.preventDefault();
			return;
		}
		const event = isJetpackCloud()
```

fushar replied:
I think what @p-jackson is trying to achieve can be done by simply adding the `busy` prop to the `<Button />`. I just tested it, it will make the link non-clickable and will not fire the`onClick` handler.

E.g.

```
<Button
  busy={ ! onboardingUrl }
  borderless
  ...
```

What do you think / could you confirm it?

p-jackson replied:
TIL about the `busy` prop. This looks like it's exactly what we need 👍

roo2 replied:
nice that works really well, thanks

fushar commented on client/components/site-selector/use-onboarding-url.ts:
```diff
@@ -0,0 +1,35 @@
+import { useExperiment } from 'calypso/lib/explat';
+import isJetpackCloud from 'calypso/lib/jetpack/is-jetpack-cloud';
+
+const GUTENBOARDING_LOCALES = [ 'en', 'en-gb' ];
```
Maybe I'm missing some context here, but can we reuse this instead?

https://github.com/Automattic/wp-calypso/blob/c7e24c921d38445d6814f165d1bd8ca26cfe75fa/packages/i18n-utils/src/locales.ts#L25

p-jackson replied:
Technically there's no reason why the Gutenboarding locales have to be English. If we found during testing that French performed just as well as English then it could be `[ 'en', 'en-gb', 'fr' ]`. Which is a reason to use our own special array: `englishLocales` and `GUTENBOARDING_LOCALES` are semantically different.

But for all intents and purposes I think the decision about which locales go to Gutenboarding _was_ made on their "english-ness". So yeah, I like the idea of using the pre-defined `englishLocales` variable.

roo2 replied:
I think there's an argument for either, I implemented it this way because I coppied what's in https://github.com/Automattic/wp-calypso/blob/trunk/client/state/selectors/get-onboarding-url.js

p-jackson reviewed (commented):
> There are other links to create a new site, but I don't believe these need to be updated (I got a bit lost in the weeds trying to do that) I don't think it's necessary to update all links.

In general, I agree with this. I think we only need to run the test on the bulk of users, not every single one. I did find one other "Add new site" button that I think is a "must do" though.

<img width="553" alt="Screenshot 2022-02-04 at 12 10 37 PM" src="https://user-images.githubusercontent.com/1500769/152444895-08eb7758-1fcc-4b12-b693-8b716ba9c326.png">

If the user only has _one_ site then the "Add new site" button isn't hidden behind the sidebar site switcher, and it turns out it's a different component because it didn't cause the `useExperiment()` hook to get called in my experiment.

p-jackson commented:
I'm also having trouble testing it with Abacus. Even when I manually assign users to a variation, `experimentAssignment?.variationName` is always `null`.

Surely the Abacus experiment doesn't have to be deployed before we can test this?

roo2 committed:
add loading state to add new site button

roo2 committed:
add experiment to unified sidebar

roo2 commented:
> I did find one other "Add new site" button that I think is a "must do" though 

I updated to add this case as well, thanks!

I also added the 'busy' state as @fushar suggested 👍, for the new case, it wasn't possible to use `busy` so I just delay rendering that link until the onboardingUrl becomes vailable

> Even when I manually assign users to a variation, experimentAssignment?.variationName is always null.

Oh hmm yes I haven't done this before, I'll try to work out if there's another way to test it before launching

roo2 commented on client/my-sites/sidebar-unified/add-new-site.jsx:
```diff
@@ -1,20 +1,22 @@
-	const onboardingUrl = useSelector( getOnboardingUrl );
-
+	if ( ! onboardingUrl ) {
+		return null;
```
I couldn't use the "busy" property here, so I think the best is just to delay rendering until onboardingUrl becomes available.

jessie-ross commented on client/components/site-selector/use-onboarding-url.ts:
```diff
@@ -0,0 +1,35 @@
+ *
+ * @returns {string}  URL of the onboarding flow for existing users.
+ */
+const useOnboardingUrl = (): string => {
```
Seems a bit strange to me that we are assigning an experiment when we are generating a URL.

I think what might be useful here is a redirect experiment - if the user navigates to `config( 'gutenboarding_url' );` we check experiment eligibility and redirect to `config( 'signup_url' )` if treatment.

arthur791004 replied:
In this way, when the user navigates to `config('gutenboarding_url')`, they may see the loading screen and wait for the experiment result. After the experiment result is ready, they'll redirect to `config( 'signup_url' )` and might need to wait for loading again.

Besides, if we redirect to `config( 'gutenboarding_url' )` first and redirect to `config( 'signup_url' )` later, calypso will load unnessasry resources for `config( 'gutenboarding_url' )` and it's not good for performance. Also, we need to take care about the page view of `config( 'gutenboarding_url' )` or something else to avoid any redundant `config( 'gutenboarding_url' )` page.

Therefore, I prefer the current way as I think it makes it simple and provides a better user experience.

razvanpapadopol replied:
Good points, @arthur791004 🙌🏼 

> a better user experience.

Not only that (because we wait for the experiment assignment to load) but also the accuracy of the experiment can be impacted by such redirect. From the previous attempts at running `/new` redirect experiments in signup controller, we've identified 2 main issues: 
* Even worse user experience for `treatment` (it's a completely different entrypoint that needs to be loaded) than for `control` depending on the connection speed and/or device. 
* Bounce rate may impact more the conversion rate for `control`. Even if the bot traffic is filtered, there are still real users that don't wait for the page to load. In this case, the exposure event (`calypso_signup_step_start`) may not even fire for `treatment` but it will fire for `control` that is immediately rendered.

These two may cancel each other sometimes but it's impossible to quantify how much.

However, if we really need to do a redirect experiment, I think there should be a way to preload the assets for both entrypoints before rendering the variation assigned for the user. It will be an even poorer user experience but at least we reduce the margin of error as much as possible.

roo2 replied:
hmm I do see @jessie-ross's point though, since basically every user who loads calypso will call useOnboardingUrl , they will be assigned to an experiment group. But how many of them are going to actually use this link in the UI and how many of them are going to load the `/new` or `/start` flow from some other link? But then we don't want to slow down the user experience by adding a redirect, and especially, we don't want one variation to be significantly slowed down by a redirect while the other isn't.

I think that in this case, we are very motivated to deactivate the /new flow for maintainance and tech strategy reasons, so perhaps running an extremely clean experiment isn't as important as it would be if this were an experimental flow that we were testing. Knowing that, my preference is to go with the current approach, and acknowledge that some users may be incorrectly assigned which will cause noise in the experiment, with the tradeoff that performance is maintained.

jessie-ross commented on client/components/site-selector/use-onboarding-url.ts:
```diff
@@ -0,0 +1,35 @@
+ */
+const useOnboardingUrl = (): string => {
+	const userLocale = useLocale();
+	const [ isLoadingExperimentAssignment, experimentAssignment ] = useExperiment(
+		'calypso_site_create_existing_users_to_start_flow'
+	);
```
We should be using assignment eligibility to filter in-code:
```suggestion
	const [ isLoadingExperimentAssignment, experimentAssignment ] = useExperiment(
		'calypso_site_create_existing_users_to_start_flow',
		{ isEligible: ! isJetpackCloud() && GUTENBOARDING_LOCALES.includes( userLocale ) },
	);
```

roo2 replied:
good catch 👍

arthur791004 reviewed (commented):
Should we also need to apply the changes to `calypso/my-sites/sidebar`? If 
there is `disable-nav-unification` in query param or the site is Jetpack sites below 9.8, the `isNavUnificationEnabled()` returns false and the calypso will use `calypso/my-sites/sidebar` rather than `calypso/my-sites/sidebar-unified`

See https://github.com/Automattic/wp-calypso/blob/9cf831eeb055e5a65784011fe1337be388658281/client/my-sites/navigation/index.jsx#L26

roo2 commented:
> Should we also need to apply the changes to calypso/my-sites/sidebar? If
there is disable-nav-unification in query param or the site is Jetpack sites below 9.8,

@arthur791004 good point, but It looks like that uses the same `SiteSelectorAddSite` component which checks using useOnboardingUrl already 👍

roo2 committed:
use eligibility criteria

arthur791004 commented:
@roo2 But I saw it uses `this.props.onboardingUrl` [here](https://github.com/Automattic/wp-calypso/blob/9cf831eeb055e5a65784011fe1337be388658281/client/my-sites/sidebar/index.jsx#L857) and the value comes from `getOnboardingUrl` selector 🤔

roo2 commented:
@arthur791004 I tested with `?disable-nav-unification` and it was still showing the `SiteSelectorAddSite` component. I actually can't find where the add site button defined in `my-sites/sidebar/index.jsx` is used
<img width="971" alt="Screen Shot 2022-02-14 at 9 33 42 pm" src="https://user-images.githubusercontent.com/22446385/153857579-3a33c7c4-1a97-4ab2-a817-dad13159131a.png">

arthur791004 commented:
@roo2 It won't show when the user has at least 2 sites. But if you change to another account with only 1 site, you'll see it

![image](https://user-images.githubusercontent.com/13596067/153984104-bf106cfa-ef5e-475c-986e-7a4b65320b84.png)

roo2 committed:
Add onboarding url to 'add site' button in sidebar

roo2 commented:
ahh, thanks for clarifying @arthur791004. I updated the sidebar to get the "add new site" url from the useOnboardingUrl function as well. This affects the case when nav unification is disabled and the user has only one site.

arthur791004 reviewed (approved):
LGTM, let' go 🚢

autumnfjeld commented:
@razvanpapadopol   We want to launch this experiment so  @roo2  will go ahead and merge.   Just wanted to double check in case there were any outstanding issues.
